libavcodec/vaapi_encode: Change the way to call async to increase

wenbinc-Bin opened the pull request:
performance

Fix: #7706. After commit 5fdcf85bbffe7451c2, vaapi encoder's performance
decrease. The reason is that vaRenderPicture() and vaSyncSurface() are
called at the same time (vaRenderPicture() always followed by a
vaSyncSurface()). Now I changed them to be called in a
asynchronous way, which will make better use of hardware.

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavcodec/vaapi_encode: Change the way to call async to increase performance

Fix: #7706. After commit 5fdcf85bbffe7451c2, vaapi encoder's performance
decrease. The reason is that vaRenderPicture() and vaSyncSurface() are
called at the same time (vaRenderPicture() always followed by a
vaSyncSurface()). When we encode stream with B frames, we need buffer to
reorder frames, so we can send serveral frames to HW at once to increase
performance. Now I changed them to be called in a
asynchronous way, which will make better use of hardware.
1080p transcoding increases about 17% fps on my environment.

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>

wenbinc-Bin commented:
On my side, Transcoding 1080p h264 on i9-9900k increase from 240fps to 270fps.

feiwan1 commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,20 +1179,34 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
-    if (err < 0)
-        return err;
-    av_assert0(pic);
+    if (av_fifo_size(ctx->encode_fifo) == 0) {
```
Any  possible to remove this check? Or check fifo size inside while with "if (av_fifo_size(ctx->encode_fifo) == MAX_ALLOC_SIZE) break;". With your logic, if fifo size is 0, and the next pic will write to fifo, and then get it out at once, this meets the case in your commit log. If change like this, you may consider if need to enlarge fifo size in init function.

wenbinc-Bin replied:
Thanks, I try to improve code like you said.

wenbinc-Bin replied:
Done, please check.

feiwan1 commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,20 +1179,31 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
+        return err;
+
+    av_fifo_generic_read(ctx->encode_fifo, &pic, sizeof(pic), NULL);
+    ctx->encode_order = pic->encode_order+1;
```
"encoder_order+1" -> "encoder_order + 1".

wenbinc-Bin replied:
done

feiwan1 commented on libavcodec/vaapi_encode.c:
```diff
@@ -2520,6 +2533,10 @@ av_cold int ff_vaapi_encode_init(AVCodecContext *avctx)
         }
     }
 
+    ctx->encode_fifo = av_fifo_alloc((MAX_PICTURE_REFERENCES+1)* sizeof(VAAPIEncodePicture *));
```
"MAX_PICTURE_REFERENCES+1)*" -> "MAX_PICTURE_REFERENCES + 1) *". You can use tools/patchcheck to scan your patch, it can detect this kind of style issue. Others looks good to me.

wenbinc-Bin replied:
done

feiwan1 commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,20 +1179,31 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
-    if (err < 0)
-        return err;
-    av_assert0(pic);
+    while (av_fifo_size(ctx->encode_fifo) <= MAX_PICTURE_REFERENCES*sizeof(VAAPIEncodePicture *)) {
```
" MAX_PICTURE_REFERENCES*sizeof" -> " MAX_PICTURE_REFERENCES * sizeof"

wenbinc-Bin replied:
done

wenbinc-Bin commented:
Any other comments?

wenbinc-Bin commented:
Will submit to upstream if there is no other comment.

xuguangxin reviewed (commented):
Will this introduce frame delay for IPPPPP pattern?

xuguangxin commented:
Is it possible to use https://github.com/intel/libva/blob/master/va/va.h#L4193 to query buffer status?
the logical maybe:
1. vaapi_encode_issue
2. push to fifo
3. if fifo is full, vaSyncBuffer with infinitely.  pop and push it to downstream. 
4. vaSyncBuffer the entire fifo with timeout 0. stop if we find a VA_STATUS_ERROR_TIMEDOUT

wenbinc-Bin commented:
@xuguangxin, Ok I will try this to improve performance.

wenbinc-Bin committed:
libavcodec/vaapi_encode: Add new API adaption and async_depth to vaapi_encoder

Add async_depth to increase encoder's performance. Reuse encode_fifo as
async buffer. Encoder puts all reordered frame to HW and then check
fifo size. If fifo < async_depth and the top frame is not ready, it will
return AVERROR(EAGAIN) to require more frames.

1080p transcoding (no B frames) with -async_depth=4 can increase 20%
performance on my environment.
The async increases performance but also introduces frame delay.

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>

xuguangxin commented on libavcodec/vaapi_encode_vp9.c:
```diff
@@ -260,6 +260,13 @@ static const AVOption vaapi_encode_vp9_options[] = {
     { "loop_filter_sharpness", "Loop filter sharpness",
       OFFSET(loop_filter_sharpness), AV_OPT_TYPE_INT, { .i64 = 4 }, 0, 15, FLAGS },
+
+#if VA_CHECK_VERSION(1, 9, 0)
```
could we add this to VAAPI_ENCODE_COMMON_OPTIONS

wenbinc-Bin replied:
I tried but #if cannot be added into macro definition.

xuguangxin replied:
you can use vaQuerySurface for < 1.9

wenbinc-Bin replied:
done

xuguangxin commented on libavcodec/vaapi_encode.c:
```diff
@@ -2520,6 +2550,13 @@ av_cold int ff_vaapi_encode_init(AVCodecContext *avctx)
         }
     }
 
+    ctx->encode_fifo = av_fifo_alloc((ctx->async_depth > MAX_PICTURE_REFERENCES ?
```
1. If we have this, How to handle low delay encode? 
2. could we adjust async_depth in one place?

wenbinc-Bin replied:
1.The delay is caused by async_depth not by the size of encode_fifo.
I create encode_fifo with at least size 3 to make sure we get high performance when encode b frames. This will not cause delay.
2. if I use vaQuerySurface when < 1.9.0, I can move async_depth option to VAAPI_ENCODE_COMMON_OPTIONS, then we can adjust this variable through command line option.

xuguangxin commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,19 +1180,46 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
+    if (!av_fifo_size(ctx->encode_fifo))
         return err;
+
+#if VA_CHECK_VERSION(1, 9, 0)
```
if < 1.9.0, we can use vaQuerySurface

wenbinc-Bin replied:
Thanks for the information, I will try.

xuguangxin commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,19 +1180,46 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
+    if (av_fifo_size(ctx->encode_fifo) < ctx->async_depth * sizeof(VAAPIEncodePicture *) &&
+        !ctx->end_of_stream) {
+        av_fifo_generic_peek(ctx->encode_fifo, &pic, sizeof(pic), NULL);
+        err = vaSyncBuffer(ctx->hwctx->display, pic->output_buffer, 0);
```
we'd better output all ready buffers. not only the current buffer, it will reduce the delay

xuguangxin commented on libavcodec/vaapi_encode.c:
```diff
@@ -150,12 +150,21 @@ static int vaapi_encode_wait(AVCodecContext *avctx,
            "(input surface %#x).\n", pic->display_order,
            pic->encode_order, pic->input_surface);
 
+#if VA_CHECK_VERSION(1, 9, 0)
```
if vaSyncBuffer returns unimplemented we may need fallback to vaSyncSurface.

wenbinc-Bin replied:
Oh, that is why pre-check fail in i965 test.
Thanks

xuguangxin commented on libavcodec/vaapi_encode.h:
```diff
@@ -455,7 +462,10 @@ int ff_vaapi_encode_close(AVCodecContext *avctx);
       OFFSET(common.desired_b_depth), AV_OPT_TYPE_INT, \
-      { .i64 = 1 }, 1, INT_MAX, FLAGS }
+      { .i64 = 1 }, 1, INT_MAX, FLAGS }, \
+    { "async_depth", "Maximum processing parallelism", \
```
better say, increase this to improve single channel performance

wenbinc-Bin replied:
done

xuguangxin commented on libavcodec/vaapi_encode.c:
```diff
@@ -1177,20 +1189,60 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
+                   "%d (%s).\n", err, vaErrorStr(err));
+            return AVERROR(EIO);
+        }
+#elif
```
we can add a timeout to vaapi_encode_wait. and call vaapi_encode_wait(..., timeout = 0) here

wenbinc-Bin replied:
Good idea, will change.
Thanks

wenbinc-Bin committed:
libavcodec/vaapi_encode: Add new API adaption to vaapi_encode

Add vaSyncBuffer to VAAPI encoder. Old version API vaSyncSurface wait
surface to complete. When surface is used for multiple operation, it
wait all operation to finish. vaSyncBuffer only wait one channel to
finish.

Add wait param to vaapi_encode_wait() to prepare for the async_depth
option. "wait=1" means wait until operation ready. "wait=0" means
query operation's status. If ready return 0, if still in progress
return EAGAIN.

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>

wenbinc-Bin commented:
submit to https://patchwork.ffmpeg.org/project/ffmpeg/patch/20211027085705.4114165-1-wenbin.chen@intel.com/
